### ISSUE TITLE
Warn developers if `main` is running for very long without producing a frame.

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#133289]
Multiple application teams have been observed to do a pattern like the following:

```dart
void main() async {
  final foo = await _setupFoo();
  PlatformDispatcher.instance.onError = mySpecialErrorHandler;
  doBar(await _someFutureValue());
  await _theNextThing(foo);
  await _anotherThingThatWeAddedLater();
  // Many more such awaits

  runApp(MaterialApp(home: MyRootWidget(specialThing: await makeSpecialThing()));
}
```

As this kind of code grows and more `await`s are added, it gets harder and harder to disentangle. But the time the product is getting ready to ship, the application is spending some significant amount of time in `main` before ever trying to render a frame.

We should print a warning about this. The only time this would be desirable is in the case where a developer is pre-warming the engine while showing non-Flutter UI, in which case they could ignore the warning.

This would be similar to the way that Android warns you if the choreographer gets delayed by too long and tells you that you "may be doing too much work on the main thread." The goal would be to have an earlier signal to the application development team to start cleaning this up before it becomes even more entangled.

@jiahaog WDYT?

**Comment by jiahaog** [comment]
Yeah I think this could be helpful. I wonder what will be a good duration before the warning is shown though. For example `await Firebase.initializeApp();` is quite ubiquitous and this can take ~100ms at least on Android.

**Comment by dnfield** [comment]
At what point have you seen this become a problem? What are the application teams willing to accept?

100ms for firebase seems like a lot to me but I don't really know all of what it's doing or how critical it is to finish it before rendering content.

**Comment by jiahaog** [comment]
> At what point have you seen this become a problem? What are the application teams willing to accept?

I don't know if there's a hard number here. Most of the time, we're already late to the party and teams have already "accepted" this delay without knowing about it. It's also not that easy to tell how long this delay is even with devtools.

Perhaps we could start off by adding a trace from the time when we run the Dart main till when the first frame is drawn? This could provide more visibility when developers use devtools as well. We could also have a set of "warnings" in devtools that looks for these things and warns the user.
